File: core/src/mindustry/client/utils/Crypto.java
package mindustry.client.utils;
/*
 * Copyright (c) 1997, 2017, Oracle and/or its affiliates. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *   - Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   - Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *
 *   - Neither the name of Oracle nor the names of its
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
 * IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import arc.math.*;
import arc.struct.*;
import org.bouncycastle.jce.provider.*;

import java.nio.charset.*;
import java.security.*;
import java.security.spec.*;
import java.util.*;
import javax.crypto.*;
import javax.crypto.spec.*;
import javax.crypto.interfaces.*;

public class Crypto{
    private KeyPair keyPair;
    public State state;
    private KeyAgreement agreement;
    private SecretKeySpec aesKey;
    private IvParameterSpec iv;
    private Cipher cipher;
    public Array<String> keyAccumulator;
    public boolean isInitializing;

    public Crypto(boolean initializing) {
        this.isInitializing = initializing;
        if(initializing) {
            state = State.WAITING_FOR_RESPONSE;
            try{
                KeyPairGenerator dhKeyPairGen = KeyPairGenerator.getInstance("DH");
                dhKeyPairGen.initialize(512);
                keyPair = dhKeyPairGen.generateKeyPair();

                agreement = KeyAgreement.getInstance("DH");
                agreement.init(keyPair.getPrivate());
            } catch(NoSuchAlgorithmException | InvalidKeyException error) {
                error.printStackTrace();
            }
        }else{
            state = State.WAITING_FOR_KEY;
        }
    }

    public byte[] getDhKey() {
        return keyPair.getPublic().getEncoded();
    }

    public boolean isReady() {
        return state == State.READY;
    }

    public byte[] respondToKey(byte[] key){
        try{
            KeyFactory keyFactory = KeyFactory.getInstance("DH");
            X509EncodedKeySpec x509KeySpec = new X509EncodedKeySpec(key);

            PublicKey incomingKey = keyFactory.generatePublic(x509KeySpec);

            /*
             * Bob gets the DH parameters associated with Alice's public key.
             * He must use the same parameters when he generates his own key
             * pair.
             */
            DHParameterSpec dhParamFromIncomingKey = ((DHPublicKey)incomingKey).getParams();

            // Bob creates his own DH key pair
            KeyPairGenerator outgoingKeyGen = KeyPairGenerator.getInstance("DH");
            outgoingKeyGen.initialize(dhParamFromIncomingKey);
            KeyPair outgoingKeyPair = outgoingKeyGen.generateKeyPair();

            // Bob creates and initializes his DH KeyAgreement object
            agreement = KeyAgreement.getInstance("DH");

            agreement.init(outgoingKeyPair.getPrivate());
            agreement.doPhase(incomingKey, true);

            return outgoingKeyPair.getPublic().getEncoded();
        } catch(NoSuchAlgorithmException | InvalidKeySpecException | InvalidAlgorithmParameterException | InvalidKeyException error) {
            error.printStackTrace();
            return null;
        }
    }

    public void receiveDhKey(byte[] key) {
        try{
            KeyFactory keyFactory = KeyFactory.getInstance("DH");
            X509EncodedKeySpec x509KeySpec = new X509EncodedKeySpec(key);
            PublicKey incomingKey = keyFactory.generatePublic(x509KeySpec);
            agreement.doPhase(incomingKey, true);
        } catch(NoSuchAlgorithmException | InvalidKeySpecException | InvalidKeyException error) {
            error.printStackTrace();
        }
    }

    public void generateAesKey() {
        try{
            byte[] sharedSecret = agreement.generateSecret();

            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            digest.update(sharedSecret);
            byte[] keyBytes = new byte[16];
            System.arraycopy(digest.digest(), 0, keyBytes, 0, keyBytes.length);
            byte[] ivBytes = new byte[16];
            System.arraycopy(digest.digest(), 16, ivBytes, 0, keyBytes.length);
            aesKey = new SecretKeySpec(keyBytes, "AES");
            iv = new IvParameterSpec(ivBytes);

            cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            state = State.READY;
        } catch(NoSuchAlgorithmException | NoSuchPaddingException error) {
            error.printStackTrace();
        }
    }

    public static void main(String[] argv) throws Exception {
        Security.addProvider(new BouncyCastleProvider());

//        /*
//         * Alice creates her own DH key pair with 2048-bit key size
//         */
//        System.out.println("ALICE: Generate DH keypair ...");
//        KeyPairGenerator aliceKpairGen = KeyPairGenerator.getInstance("DH");
//        aliceKpairGen.initialize(2048);
//        KeyPair aliceKpair = aliceKpairGen.generateKeyPair();
//
//        // Alice creates and initializes her DH KeyAgreement object
//        System.out.println("ALICE: Initialization ...");
//        KeyAgreement aliceKeyAgree = KeyAgreement.getInstance("DH");
//        aliceKeyAgree.init(aliceKpair.getPrivate());
//
//        // Alice encodes her public key, and sends it over to Bob.
//        byte[] alicePubKeyEnc = aliceKpair.getPublic().getEncoded();
//
//        /*
//         * Let's turn over to Bob. Bob has received Alice's public key
//         * in encoded format.
//         * He instantiates a DH public key from the encoded key material.
//         */
//        KeyFactory bobKeyFac = KeyFactory.getInstance("DH");
//        X509EncodedKeySpec x509KeySpec = new X509EncodedKeySpec(alicePubKeyEnc);
//
//        PublicKey alicePubKey = bobKeyFac.generatePublic(x509KeySpec);
//
//        /*
//         * Bob gets the DH parameters associated with Alice's public key.
//         * He must use the same parameters when he generates his own key
//         * pair.
//         */
//        DHParameterSpec dhParamFromAlicePubKey = ((DHPublicKey)alicePubKey).getParams();
//
//        // Bob creates his own DH key pair
//        System.out.println("BOB: Generate DH keypair ...");
//        KeyPairGenerator bobKpairGen = KeyPairGenerator.getInstance("DH");
//        bobKpairGen.initialize(dhParamFromAlicePubKey);
//        KeyPair bobKpair = bobKpairGen.generateKeyPair();
//
//        // Bob creates and initializes his DH KeyAgreement object
//        System.out.println("BOB: Initialization ...");
//        KeyAgreement bobKeyAgree = KeyAgreement.getInstance("DH");
//        bobKeyAgree.init(bobKpair.getPrivate());
//
//        // Bob encodes his public key, and sends it over to Alice.
//        byte[] bobPubKeyEnc = bobKpair.getPublic().getEncoded();
//
//        /*
//         * Alice uses Bob's public key for the first (and only) phase
//         * of her version of the DH
//         * protocol.
//         * Before she can do so, she has to instantiate a DH public key
//         * from Bob's encoded key material.
//         */
//        KeyFactory aliceKeyFac = KeyFactory.getInstance("DH");
//        x509KeySpec = new X509EncodedKeySpec(bobPubKeyEnc);
//        PublicKey bobPubKey = aliceKeyFac.generatePublic(x509KeySpec);
//        System.out.println("ALICE: Execute PHASE1 ...");
//        aliceKeyAgree.doPhase(bobPubKey, true);
//
//        /*
//         * Bob uses Alice's public key for the first (and only) phase
//         * of his version of the DH
//         * protocol.
//         */
//        System.out.println("BOB: Execute PHASE1 ...");
//        bobKeyAgree.doPhase(alicePubKey, true);
//
//        /*
//         * At this stage, both Alice and Bob have completed the DH key
//         * agreement protocol.
//         * Both generate the (same) shared secret.
//         */
//        byte[] aliceSharedSecret = aliceKeyAgree.generateSecret();
//        int aliceLen = aliceSharedSecret.length;
//        byte[] bobSharedSecret = new byte[aliceLen];
//        int bobLen = bobKeyAgree.generateSecret(bobSharedSecret, 0);
//        System.out.println("Alice secret: " +
//        toHexString(aliceSharedSecret));
//        System.out.println("Bob secret: " +
//        toHexString(bobSharedSecret));
//        if (!java.util.Arrays.equals(aliceSharedSecret, bobSharedSecret))
//            throw new Exception("Shared secrets differ");
//        System.out.println("Shared secrets are the same");
//
//        /*
//         * Now let's create a SecretKey object using the shared secret
//         * and use it for encryption. First, we generate SecretKeys for the
//         * "AES" algorithm (based on the raw shared secret data) and
//         * Then we use AES in CBC mode, which requires an initialization
//         * vector (IV) parameter. Note that you have to use the same IV
//         * for encryption and decryption: If you use a different IV for
//         * decryption than you used for encryption, decryption will fail.
//         *
//         * If you do not specify an IV when you initialize the Cipher
//         * object for encryption, the underlying implementation will generate
//         * a random one, which you have to retrieve using the
//         * javax.crypto.Cipher.getParameters() method, which returns an
//         * instance of java.security.AlgorithmParameters. You need to transfer
//         * the contents of that object (e.g., in encoded format, obtained via
//         * the AlgorithmParameters.getEncoded() method) to the party who will
//         * do the decryption. When initializing the Cipher for decryption,
//         * the (reinstantiated) AlgorithmParameters object must be explicitly
//         * passed to the Cipher.init() method.
//         */
//        System.out.println("Use shared secret as SecretKey object ...");
//        SecretKeySpec bobAesKey = new SecretKeySpec(bobSharedSecret, 0, 16, "AES");
//        SecretKeySpec aliceAesKey = new SecretKeySpec(aliceSharedSecret, 0, 16, "AES");
//
//        /*
//         * Bob encrypts, using AES in CBC mode
//         */
//        Cipher bobCipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
//        bobCipher.init(Cipher.ENCRYPT_MODE, bobAesKey);
//        byte[] cleartext = "This is just an example".getBytes();
//        byte[] ciphertext = bobCipher.doFinal(cleartext);
//
//        // Retrieve the parameter that was used, and transfer it to Alice in
//        // encoded format
//        byte[] encodedParams = bobCipher.getParameters().getEncoded();
//
//        /*
//         * Alice decrypts, using AES in CBC mode
//         */
//
//        // Instantiate AlgorithmParameters object from parameter encoding
//        // obtained from Bob
//        AlgorithmParameters aesParams = AlgorithmParameters.getInstance("AES");
//        aesParams.init(encodedParams);
//        Cipher aliceCipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
//        aliceCipher.init(Cipher.DECRYPT_MODE, aliceAesKey, aesParams);
//        byte[] recovered = aliceCipher.doFinal(ciphertext);
//        if (!java.util.Arrays.equals(cleartext, recovered))
//            throw new Exception("AES in CBC mode recovered text is " +
//            "different from cleartext");
//        System.out.println("AES in CBC mode recovered text is same as cleartext");
//        Crypto client0 = new Crypto(true);
//        Crypto client1 = new Crypto(false);
//        byte[] firstKey = client0.getDhKey();
//        byte[] secondKey = client1.respondToKey(firstKey);
//        client0.receiveDhKey(secondKey);
//
//        client0.generateAesKey();
//        client1.generateAesKey();
//
//        System.out.println(client0.state);
//        System.out.println(client1.state);
//
//        byte[] input = new byte[4096];
//        new Random().nextBytes(input);
//
//        byte[] encrypted = client0.encrypt(input);
//        System.out.println(Arrays.toString(encrypted));
//        System.out.println(Arrays.toString(input));
//        byte[] decrypted = client1.decrypt(encrypted);
//        System.out.println(Arrays.toString(decrypted));
//        System.out.println(Arrays.equals(input, decrypted));
//
//        String encryptedText = client0.encryptString("Hello!");
//        System.out.println(new String(client1.decrypt(encryptedText), StandardCharsets.UTF_8));


        Crypto client0 = new Crypto(true);
        Crypto client1 = new Crypto(false);

        Array<String> message = client0.getKey();
        Array<String> messages2 = client1.fromMessages(message);
        client0.fromMessages(messages2);
        client0.generateAesKey();
        client1.generateAesKey();
        System.out.println(client0.encryptString("AAAAAAAAA"));
        System.out.println(new String(client1.decrypt(client0.encryptString("AAAAAAAAA")), StandardCharsets.UTF_8));
    }

    public Array<String> getKey(){
        String key = Base256Coder.encode(getDhKey());
        Array<String> parts = new Array<>();
        for (int i = 0; i < key.length(); i += 100) {
            parts.add(key.substring(i, Math.min(i + 100, key.length())));
        }
        return parts;
    }

    public Array<String> fromMessages(Array<String> inp){
        StringBuilder builder = new StringBuilder();
        for(String s : inp){
            builder.append(s);
        }

        if(state == State.WAITING_FOR_KEY){
            Array<String> parts = new Array<>();
            String key = Base256Coder.encode(respondToKey(Base256Coder.decode(builder.toString())));
            for (int i = 0; i < key.length(); i += 100) {
                parts.add(key.substring(i, Math.min(i + 100, key.length())));
            }
            return parts;
        }else{
            receiveDhKey(Base256Coder.decode(builder.toString()));
            return null;
        }
    }

    public byte[] encrypt(byte[] input) {
        try{
            cipher.init(Cipher.ENCRYPT_MODE, aesKey, iv);
            return cipher.doFinal(input);
        } catch(InvalidKeyException | IllegalBlockSizeException | BadPaddingException | InvalidAlgorithmParameterException error) {
            error.printStackTrace();
            return null;
        }
    }

    public byte[] decrypt(byte[] input) {
        try{
            cipher.init(Cipher.DECRYPT_MODE, aesKey, iv);
            return cipher.doFinal(input);
        } catch(InvalidKeyException | IllegalBlockSizeException | BadPaddingException | InvalidAlgorithmParameterException error) {
            error.printStackTrace();
            return null;
        }
    }

    public String encryptString(byte[] input) {
        return Base256Coder.encode(encrypt(input));
    }

    public String encryptString(String input) {
        return Base256Coder.encode(encrypt(input.getBytes(StandardCharsets.UTF_8)));
    }

    public byte[] decrypt(String input) {
        return decrypt(Base256Coder.decode(input));
    }

    /*
     * Converts a byte to hex digit and writes to the supplied buffer
     */
    private static void byte2hex(byte b, StringBuffer buf) {
        char[] hexChars = { '0', '1', '2', '3', '4', '5', '6', '7', '8',
        '9', 'A', 'B', 'C', 'D', 'E', 'F' };
        int high = ((b & 0xf0) >> 4);
        int low = (b & 0x0f);
        buf.append(hexChars[high]);
        buf.append(hexChars[low]);
    }

    /*
     * Converts a byte array to hex string
     */
    private static String toHexString(byte[] block) {
        StringBuffer buf = new StringBuffer();
        int len = block.length;
        for (int i = 0; i < len; i++) {
            byte2hex(block[i], buf);
            if (i < len-1) {
                buf.append(":");
            }
        }
        return buf.toString();
    }

    @Override
    public String toString(){
        return "Client state: " + state + " ready: " + isReady();
    }
}

enum State {
    WAITING_FOR_KEY, WAITING_FOR_RESPONSE, READY;
}

File: core/src/mindustry/client/crypto/Crypto.kt
package mindustry.client.crypto

import mindustry.client.Initializable
import org.bouncycastle.crypto.agreement.X25519Agreement
import org.bouncycastle.crypto.generators.Ed25519KeyPairGenerator
import org.bouncycastle.crypto.generators.X25519KeyPairGenerator
import org.bouncycastle.crypto.params.*
import org.bouncycastle.crypto.signers.Ed25519Signer
import org.bouncycastle.jce.provider.BouncyCastleProvider
import org.bouncycastle.math.ec.rfc8032.Ed25519
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.security.SecureRandom
import java.security.Security
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec
import kotlin.random.Random

typealias XPublicKey = X25519PublicKeyParameters
typealias XPrivateKey = X25519PrivateKeyParameters

typealias EdPublicKey = Ed25519PublicKeyParameters
typealias EdPrivateKey = Ed25519PrivateKeyParameters

/**
 * Utility object that handles signing and encryption.
 * Uses ED25519 for signing and X25519 to generate a shared AES key.
 * DISCLAIMER: I am not a cryptographer.  This implementation may be buggy.
 */
object Crypto : Initializable {
    /** The length of a signature, in bytes. */
    var signatureSize = Ed25519.SIGNATURE_SIZE
    private lateinit var signatureEngine: Ed25519Signer
    private lateinit var aes: Cipher
    /** The secure random number generator. */
    private val random = SecureRandom.getInstanceStrong()

    /** Initializes cryptography stuff, must be called before usage. */
    override fun initializeAlways() {
        Security.addProvider(BouncyCastleProvider())
        signatureEngine = Ed25519Signer()
        aes = Cipher.getInstance("AES/CBC/PKCS5Padding")
    }

    fun generateKeyQuad(): KeyQuad {
        val generator = Ed25519KeyPairGenerator()
        generator.init(Ed25519KeyGenerationParameters(random))
        val edKeys = generator.generateKeyPair()
        val xGenerator = X25519KeyPairGenerator()
        xGenerator.init(X25519KeyGenerationParameters(random))
        val xKeys = xGenerator.generateKeyPair()

        return KeyQuad(xKeys.private as XPrivateKey, xKeys.public as XPublicKey, edKeys.private as EdPrivateKey, edKeys.public as EdPublicKey)
    }

    /**
     * Signs an input with a private key (verify with [verify]).
     * Note: vulnerable to replay attack, do not use on its own.
     */
    fun sign(input: ByteArray, key: Ed25519PrivateKeyParameters): ByteArray {
        signatureEngine.init(true, key)
        signatureEngine.update(input, 0, input.size)
        return signatureEngine.generateSignature()
    }

    /** Encrypts a message with AES. */
    fun encrypt(input: ByteArray, key: SecretKeySpec, iv: IvParameterSpec): ByteArray {
        aes.init(Cipher.ENCRYPT_MODE, key, iv)
        return aes.doFinal(input)
    }

    /** Decrypts a message with AES. */
    fun decrypt(input: ByteArray, key: SecretKeySpec, iv: IvParameterSpec): ByteArray {
        aes.init(Cipher.DECRYPT_MODE, key, iv)
        return aes.doFinal(input)
    }

    /**
     * Verifies a signature from [sign] given the input, signature, and corresponding public key.
     */
    fun verify(original: ByteArray, sign: ByteArray, key: EdPublicKey): Boolean {
        signatureEngine.init(false, key)
        signatureEngine.update(original, 0, original.size)
        return signatureEngine.verifySignature(sign)
    }
}

/** Represents a client with a [KeyQuad].  It talks to exactly one other client and deals with generating the shared
 * secret and encryption. */
class CryptoClient(val key: KeyQuad) {
    var sharedSecret: ByteArray = byteArrayOf()
    lateinit var aesKey: SecretKeySpec
    lateinit var iv: IvParameterSpec
    lateinit var otherKey: EdPublicKey

    /** Generates the shared secret and initializes the AES cipher.  Must be called before using the other methods. */
    fun generate(other: PublicKeyPair) {
        otherKey = other.edPublicKey
        val agreement = X25519Agreement()
        agreement.init(key.xPrivateKey)
        sharedSecret = ByteArray(agreement.agreementSize)
        agreement.calculateAgreement(other.xPublicKey, sharedSecret, 0)

        val digest = MessageDigest.getInstance("SHA-256")
        digest.update(sharedSecret)
        val keyBytes = ByteArray(16)
        System.arraycopy(digest.digest(), 0, keyBytes, 0, keyBytes.size)
        val ivBytes = ByteArray(16)
        System.arraycopy(digest.digest(), 16, ivBytes, 0, keyBytes.size)
        aesKey = SecretKeySpec(keyBytes, "AES")
        iv = IvParameterSpec(ivBytes)
    }

    fun encrypt(input: ByteArray): ByteArray {
        return Crypto.encrypt(input, aesKey, iv)
    }

    fun decrypt(input: ByteArray): ByteArray {
        return Crypto.decrypt(input, aesKey, iv)
    }

    fun sign(input: ByteArray): ByteArray {
        return Crypto.sign(input, key.edPrivateKey)
    }

    fun verify(signature: ByteArray, original: ByteArray): Boolean {
        return Crypto.verify(original, signature, otherKey)
    }
}

fun main() {
    Crypto.initializeAlways()

    // Test serialization
    val quad = Crypto.generateKeyQuad()
    val decoded = KeyQuad(quad.serialize())
    assert(quad == decoded)

    val pair = PublicKeyPair(quad)
    val decodedPair = PublicKeyPair(pair.serialize())
    assert(pair == decodedPair)

    val client1 = CryptoClient(Crypto.generateKeyQuad())
    val client2 = CryptoClient(Crypto.generateKeyQuad())

    client1.generate(PublicKeyPair(client2.key.xPublicKey, client2.key.edPublicKey))
    client2.generate(PublicKeyPair(client1.key.xPublicKey, client1.key.edPublicKey))

    assert(client1.sharedSecret.contentEquals(client2.sharedSecret))
    assert(client1.aesKey == client2.aesKey)
    assert(client1.iv.iv.contentEquals(client2.iv.iv))

    val original = Random.Default.nextBytes(1024)
    assert(client1.verify(client2.sign(original), original))
    assert(client2.verify(client1.sign(original), original))
}
